File: node-js-test/src/domain/repositories/part.repository.ts
import { BaseRepository } from 'src/shared/bases/base.repository';
import {
  IPaginatedResult,
  IPaginationParams,
} from 'src/shared/interfaces/pagination.interface';
import { PartEntity } from '../entities/part.entity';

export abstract class PartRepository extends BaseRepository<PartEntity> {
  abstract findByCompanyId(
    companyId: string,
    category?: string,
  ): Promise<PartEntity[]>;
  abstract findPageByCompanyId(
    companyId: string,
    pagination: IPaginationParams,
    category?: string,
  ): Promise<IPaginatedResult<PartEntity>>;
  abstract findByIdAndCompanyId(
    id: string,
    companyId: string,
  ): Promise<PartEntity | null>;
}

File: golang-test/internal/domain/repository/part_repository.go
package repository

import (
	"context"

	"github.com/mateh3007/karhub/golang-test/internal/domain/entity"
	"github.com/mateh3007/karhub/golang-test/internal/domain/pagination"
)

type PartRepository interface {
	Create(ctx context.Context, part *entity.Part) (*entity.Part, error)
	FindByID(ctx context.Context, id string) (*entity.Part, error)
	Update(ctx context.Context, part *entity.Part) (*entity.Part, error)
	Delete(ctx context.Context, id string) error

	// FindByCompanyID returns every part needing evaluation for restock
	// priorities, unpaginated — urgencyScore isn't a DB column, so the
	// full company-scoped set has to come back and get sorted in memory
	// (see ADR 0012 in node-js-test). FindPageByCompanyID is the one used
	// by the plain, paginated GET /parts listing.
	FindByCompanyID(ctx context.Context, companyID string) ([]*entity.Part, error)
	FindPageByCompanyID(ctx context.Context, companyID string, params pagination.Params, category *string) (pagination.Result[*entity.Part], error)
	FindByIDAndCompanyID(ctx context.Context, id, companyID string) (*entity.Part, error)
}
